File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        stack = []
        maxProfit = 0
        for p in prices[::-1]:
            if not stack:
                stack.append(p)
            else:
                if p > stack[-1]:
                    stack.append(p)
                else:
                    currProfit = stack[-1]-p
                    if maxProfit < currProfit:
                        maxProfit = currProfit
        return maxProfit

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.cpp
#include <vector>
#include <stack>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int maxProfit = 0;
        int buyPrice = prices[0];
        for (int i = 1; i < prices.size(); i++) {
            int price = prices[i];
            if (price < buyPrice)  {
                buyPrice = price;
            }
           else if (price > buyPrice && maxProfit < (price - buyPrice))    {
                maxProfit = price - buyPrice;
            }
            
        }
        return maxProfit;
    }
};